src/io/logwatcher.rs

```diff
@@ -1,121 +1,126 @@
+                log::error!("Unable to read modification time of watched file.");
+            }
+            let mut buff: Vec<u8> = Vec::new();
+            let _ = file.read_to_end(&mut buff);
 
-                // Check if file has been shortened
-                if self.reader.get_ref().metadata().unwrap().len() < self.pos {
-                    log::warn!("Console.log file was reset");
-                    self.pos = self.reader.get_ref().metadata().unwrap().len();
-                    self.last_activity = Instant::now();
-                }
+            let mut reader = Cursor::new(buff);
+            reader.set_position(self.cpos);
 
-                // Reopen the log file if nothing has happened for long enough in case the file has been replaced.
-                let time = Instant::now().duration_since(self.last_activity);
-                if time.as_secs() > 10 {
-                    log::debug!("No activity in 10 seconds, reopening.");
-                    let f = match File::open(&self.filepath) {
-                        Ok(x) => x,
-                        Err(_) => return None,
-                    };
+            if let Ok(mod_time) = file.metadata().unwrap().modified() {
+                Some(ReadMD(Some(mod_time), reader))
+            } else {
+                Some(ReadMD(None, reader)) // TODO: fix this grossness
+            }
+        } else {
+            None
+        }
+    }
 
-                    let metadata = match f.metadata() {
-                        Ok(x) => x,
-                        Err(_) => return None,
-                    };
+    fn read_lines(&mut self) {
+        if let Some(mut read_md) = self.seek_to_pos() {
+            self.last_read = read_md.0;
 
-                    let mut reader = BufReader::new(f);
-                    let pos = metadata.len();
-                    reader.seek(SeekFrom::Start(pos)).unwrap();
+            let mut data: Vec<u8> = Vec::new();
+            if let Ok(bytes_read) = read_md.1.read_to_end(&mut data) {
+                self.cpos += bytes_read as u64;
+            } else {
+                log::error!("Could not read file data!");
+                return;
+            }
 
-                    self.pos = pos;
-                    self.reader = reader;
-                    self.last_activity = Instant::now();
-                    return None;
-                }
+            let data_str = String::from_utf8_lossy(&data);
+            let mut lines: VecDeque<String> = VecDeque::new();
+            let iter = data_str.split('\n');
 
-                self.reader.seek(SeekFrom::Start(self.pos)).unwrap();
-                return None;
-            }
-            Err(err) => {
-                log::error!("Logwatcher error: {}", err);
+            for str in iter {
+                if str.is_empty() || str.trim().is_empty() {
+                    continue;
+                }
+                lines.push_back(str.to_string());
             }
+
+            self.lines_buf = lines;
```

Comment:
I still think this should be replaced with that snippet from an earlier discussion.

Reply:
but does that actually deal with lines that are just "  " or " \n" or otherwise just containing whitespace?

Reply:
I think it would be
```rust
            let data_str = String::from_utf8_lossy(&data);
            self.lines_buf.extend(data_str.lines().map(str::to_string));
```

Reply:
It splits by any `\n` or `\r\n`, but then you can also just filter the iter for only whitespace lines as well.

Reply:
So maybe something like this instead:
```rust
            let data_str = String::from_utf8_lossy(&data);
            self.lines_buf.extend(
                data_str
                    .lines()
                    .map(str::trim)
                    .filter(|s| !s.is_empty())
                    .map(str::to_string),
            );
```

Reply:
ok i turned it into this - is this necessary?
```rs
            self.lines_buf.extend(
                data_str.lines()
                .map(str::to_string)
                .filter(|x| !(x.trim().is_empty()))
            );
```

Reply:
Maybe map to string after filter.

Reply:
ah righto, was making an assumption that working with teh `&&str` the iter spits out would be gross but it isnt any different at all

Reply:
trying to see if I cant tone down the hyper::proto logging at the same time

Reply:
We can look at that in a separate MR

Reply:
yeah I think that is more something we need to dig into tappet with? not sure yet.

Reply:
No, the logging crate should provide ways of controlling the level and source of log messages. Configuring log4rs looks a little complicated though, it was easy with env_logger but that didn't provide output to a file.

Reply:
oh it lets you config logs on a module or crate level granularity? thats very nice actually